File: BasicRxJavaSample/app/src/main/java/com/kinglloy/example/android/observability/Injection.java
package com.kinglloy.example.android.observability;

import android.content.Context;

import com.kinglloy.example.android.observability.persistence.LocalUserDataSource;
import com.kinglloy.example.android.observability.persistence.UsersDatabase;
import com.kinglloy.example.android.observability.ui.ViewModelFactory;

/**
 * Enables injection of data sources.
 *
 * @author jinyalin
 * @since 2018/1/21.
 */

public class Injection {
    public static UserDataSource provideUserDataSource(Context context) {
        UsersDatabase database = UsersDatabase.getInstance(context);
        return new LocalUserDataSource(database.userDao());
    }

    public static ViewModelFactory provideViewModelFactory(Context context) {
        UserDataSource dataSource = provideUserDataSource(context);
        return new ViewModelFactory(dataSource);
    }
}

File: BasicRxJavaSampleKotlin/app/src/main/java/com/kinglloy/example/android/kotlin/Injection.kt
package com.kinglloy.example.android.kotlin

import android.content.Context
import com.kinglloy.example.android.kotlin.persistence.UserDao
import com.kinglloy.example.android.kotlin.persistence.UsersDatabase
import com.kinglloy.example.android.kotlin.ui.ViewModelFactory

/**
 * Enables injection of data sources.
 * @author jinyalin
 * @since 2018/1/22.
 */

object Injection {
    fun provideUserDataSource(context: Context): UserDao {
        val database = UsersDatabase.getInstance(context)
        return database.userDao()
    }

    fun provideViewModelFactory(context: Context): ViewModelFactory {
        val dataSource = provideUserDataSource(context)
        return ViewModelFactory(dataSource)
    }
}